interpreter/prune.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"fmt"
+	"github.com/golang/protobuf/ptypes/struct"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	expr "github.com/google/cel-spec/proto/v1/syntax"
+)
+
+type exprPruner struct {
+	expr    *expr.Expr
+	program Program
+	state   EvalState
+}
+
+func PruneExpr(expr *expr.Expr, program Program, state EvalState) {
+	pruner := &exprPruner{
+		expr:    expr,
+		program: program,
+		state:   state}
+	pruner.prune(expr)
+}
+
+func ConstantFoldExpr(expr *expr.ParsedExpr, interpreter Interpreter, container string) interface{} {
```

Comment:
Maybe provide the EvalState here from a previous run, rather than executing the interpreter in a given container.

Reply:
I listed the typical use case I thought of for PruneExpr and ConstantFoldExpr. I also noticed what we do here is related to caching function call results(especially in case of ConstantFoldExpr). I think it is important to give the users the flexibility to prepare their own runtime environment in constant fold case. They might choose how they provide overloads to enable/disable constant folding of function calls(and effectively enabling/disabling caching function call results). PruneExpr interface is the one that does what you want I think. I thought of that being used in cases where caller prunes tree right after evaluating the expression with some unknown values.